pytensor/scalar/loop.py

```diff
@@ -0,0 +1,388 @@
+
+    This Op can be used for the gradient of other Scalar Ops.
+    It is much more restricted that `Scan` in that the entire inner graph must be composed of Scalar operations.
+
+    """
+
+    init_param: Tuple[str, ...] = (
+        "init",
+        "update",
+        "constant",
+        "until",
+        "until_condition_failed",
+    )
+
+    def __init__(
+        self,
+        init: Sequence[Variable],
+        update: Sequence[Variable],
+        constant: Optional[Sequence[Variable]] = None,
+        until: Optional[Variable] = None,
+        until_condition_failed: Literal["ignore", "warn", "raise"] = "warn",
+        name="ScalarLoop",
+    ):
+        if until_condition_failed not in ["ignore", "warn", "raise"]:
+            raise ValueError(
+                f"Invalid until_condition_failed: {until_condition_failed}"
+            )
+
+        if constant is None:
+            constant = []
+        if not len(init) == len(update):
+            raise ValueError("An update must be given for each init variable")
+        if until:
+            inputs, (*outputs, until) = clone([*init, *constant], [*update, until])
+            self.outputs = copy([*outputs, until])
+        else:
+            inputs, outputs = clone([*init, *constant], update)
+            self.outputs = copy(outputs)
+        self.inputs = copy(inputs)
+
+        self.inputs_type = tuple([input.type for input in inputs])
+        self.outputs_type = tuple([output.type for output in outputs])
+        self.nin = len(inputs) + 1  # n_steps is not part of the inner graph
+        self.nout = len(outputs)  # until is not output
+        self.is_while = bool(until)
+        # TODO: Use this as part of equality and hash
+        self.until_condition_failed = until_condition_failed
+        self.name = name
+        self._validate_fgraph(FunctionGraph(self.inputs, self.outputs, clone=False))
+        super().__init__()
+
+    def output_types(self, input_types):
+        return self.outputs_type
+
+    def _validate_fgraph(self, fgraph: FunctionGraph) -> None:
+        for node in fgraph.apply_nodes:
+            if not isinstance(node.op, ScalarOp):
+                raise TypeError(
+                    "The fgraph of ScalarLoop must be composed exclusively of ScalarOp nodes"
+                )
+
+        init = fgraph.inputs
+        update = fgraph.outputs
+
+        if self.is_while:
+            *update, until = update
+            if not until.type.dtype == "bool":
+                raise TypeError(
+                    f"Until condition must be boolean, got {until}({until.type.dtype})"
+                )
+
+        for i, u in zip(init, update):
+            if i.type != u.type:
+                raise TypeError(
+                    "Init and update types must be the same: "
+                    f"{i}({i.type}) != {u}({u.type})"
+                )
+        if set(init) & set(update):
+            raise ValueError(
```

Comment:
A user could pass the same variable multiple times as `init` which would lead to problems, because the code-generating logic assumes distinct variables. I actually did that mistake when implementing some of the scalar loops